command-line-interface.md

```diff
@@ -0,0 +1,150 @@
+# Operations
+
+A conformant [runtime][] MUST provide an executable (called `funC` in the following examples).
+That executable MUST support commands with the following template:
+
+```sh
+$ funC [global-options] <COMMAND> [command-specific-options] <command-specific-arguments>
+```
+
+## Global options
+
+None are required, but the runtime MAY support options that start with at least one hyphen.
+Global options MAY take positional arguments (e.g. `--log-level debug`).
+Command names MUST NOT start with hyphens.
+The option parsing MUST be such that `funC <COMMAND>` is unambiguously an invocation of `<COMMAND>` (even for commands not specified in this document).
+If the runtime is invoked with an unrecognized command, it MUST exit with a nonzero exit code and MAY log a warning to stderr.
+
+## Character encodings
+
+This API specification does not cover character encodings, but runtimes SHOULD conform to their native operating system.
+For example, POSIX systems define [`LANG` and related environment variables][posix-lang] for [declaring][posix-locale-encoding] [locale-specific character encodings][posix-encoding], so a runtime in an `en_US.UTF-8` locale SHOULD write its [version](#version) to stdout in [UTF-8][].
+
+## Commands
+
+### version
```

Comment:
On Sun, Jul 10, 2016 at 06:45:27AM -0700, Mike Brown wrote:
> > +### version
>
> Sounds like you like the idea of supporting version via --version
> and/or version command.  Let's do that.

In order of decreasing preference for this spec:

1. Require a ‘version’ command.

2. Require a ‘--version’ global option, and explain that it is a
   command-like option and that the interaction between command-like
   options and actual commands (e.g. ‘funC --version state …’) is
   unspecified.

3. Require either 1 or 2.

4. Require both 1 and 2.

3 and 4 are at the bottom of my list, because once we have either 1 or
2 spelled out in this spec, I'd rather not complicate things by
spelling out the other.  And with 3, all consumers relying on this
spec will require a “try one and fall back to the other if the first
fails” wrapper to reliably determine the runtime version.

I'm in favor of *the runtime* supporting both 1 and 2, but this spec
is just about the minimum API we *require* for compliance testing.
And the fewer requirements we make, the better.  Here is what
compliance would look like for various runtime implementations:

* Runtime provides ‘version’: compatible with 1, 3, and 4.
* Runtime provides ‘--version’: compatible with 2, 3, and 4.
* Runtime provides both: compatible with 1, 2, 3, and 4.
* Runtime provides neither: compatible with none.